datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs

```diff
@@ -129,6 +139,7 @@ impl<S: SimplifyInfo> ExprSimplifier<S> {
         expr.rewrite(&mut const_evaluator)?
             .rewrite(&mut simplifier)?
             .rewrite(&mut or_in_list_simplifier)?
+            .rewrite(&mut guarantee_rewriter)?
             // run both passes twice to try an minimize simplifications that we missed
             .rewrite(&mut const_evaluator)?
             .rewrite(&mut simplifier)
```

Comment:
What do you think of such a loop to cover every simplification case and make it easier to accommodate future simplifications, or would it be unnecessary?
```
        loop {
            let original_expr = expr.clone();
            expr = expr
                .rewrite(&mut const_evaluator)?
                .rewrite(&mut simplifier)?
                .rewrite(&mut or_in_list_simplifier)?
                .rewrite(&mut guarantee_rewriter)?;

            if expr == original_expr {
                break;
            }
        }
```

Reply:
This is a neat idea. I think we should try it in a follow on PR.

ALso, If we did this I would also suggest adding a limit on the number of loops (to avoid a "ping-poing" infinite loop where passes rewrite an expression back and forth)